### PR TITLE
Add pipeline and trigger version

### PR DESCRIPTION
**Pull request by piyush-garg**
This will add pipeline and trigger version to the status
of the config CR

```
$ oc get config.operator.tekton.dev cluster -o json | jq '.status.conditions'
[
  {
    "code": "installed",
    "pipelineVersion": "v0.14.3",
    "triggersVersion": "v0.6.1",
    "version": "devel"
  },
  {
    "code": "applied-addons",
    "pipelineVersion": "v0.14.3",
    "triggersVersion": "v0.6.1",
    "version": "devel"
  },
  {
    "code": "validated-pipeline",
    "pipelineVersion": "v0.14.3",
    "triggersVersion": "",
    "version": "devel"
  },
  {
    "code": "applied-pipeline",
    "pipelineVersion": "",
    "triggersVersion": "",
    "version": "devel"
  }
]

```

### CODE REVIEW

File: pkg/controller/config/config_controller.go [hunk 1]
```diff
@@ -323,7 +323,19 @@ func (r *ReconcileConfig) applyAddons(req reconcile.Request, cfg *op.Config) (re
 	log.Info("successfully applied all addon resources")
 
 	err = r.updateStatus(cfg, op.ConfigCondition{Code: op.AppliedAddons, Version: flag.TektonVersion})
-	return reconcile.Result{Requeue: true}, err
+	if err != nil {
+		return reconcile.Result{}, err
+
+	}
+
+	labels := r.addons.Filter(mf.ByKind("Deployment"), mf.ByName(flag.TriggerControllerName)).Resources()[0].GetLabels()
```
Comment (nikhil-thomas) [inline pkg/controller/config/config_controller.go:331]: could you move this into a function which accepts, deploymentName and label name as args

File: pkg/controller/config/config_controller.go [hunk 2]
```diff
@@ -412,6 +424,14 @@ func (r *ReconcileConfig) validatePipeline(req reconcile.Request, cfg *op.Config
 		return reconcile.Result{}, err
 
 	}
+
+	labels := r.pipeline.Filter(mf.ByKind("Deployment"), mf.ByName(flag.PipelineControllerName)).Resources()[0].GetLabels()
+	version := labels["pipeline.tekton.dev/release"]
```
Comment (nikhil-thomas) [inline pkg/controller/config/config_controller.go:429]: same here, (let us move this into a function

**Comment by nikhil-thomas** [comment]
i see that this patch adds PipelineVersion and TriggersVersion.
Should we add ClusterTask Version as well. I understand that there is no label from where we can read this information at present.
But, I think it will be valuable for pir users

Either we can assume that there will be a label on our ClusterTasks (and pick one to read the label)
or we can take the pipeline `vmajor.minor`

cc @vdemeester

File: pkg/controller/config/config_controller.go (hunk 1 above)
Comment (piyush-garg) [reply pkg/controller/config/config_controller.go:331]: make sense :D

Comment (piyush-garg) [reply pkg/controller/config/config_controller.go]: resolved

File: pkg/controller/config/config_controller.go (hunk 2 above)
Comment (piyush-garg) [reply pkg/controller/config/config_controller.go]: done

**Comment by vdemeester** [comment]
> i see that this patch adds PipelineVersion and TriggersVersion.
> Should we add ClusterTask Version as well. I understand that there is no label from where we can read this information at present.
> But, I think it will be valuable for pir users
> 
> Either we can assume that there will be a label on our ClusterTasks (and pick one to read the label)
> or we can take the pipeline `vmajor.minor`
> 
> cc @vdemeester

What would be the `ClusterTask` version ? (for example, for 1.1 release)

**Comment by piyush-garg** [comment]
@vdemeester We don't have version label in cluster task. But we will cut the tag with the same as pipeline version v0.14.3

File: pkg/controller/config/config_controller.go [hunk 3]
```diff
@@ -412,6 +429,13 @@ func (r *ReconcileConfig) validatePipeline(req reconcile.Request, cfg *op.Config
 		return reconcile.Result{}, err
 
 	}
+
+	version := getComponentVersion(r.pipeline, flag.PipelineControllerName, "pipeline.tekton.dev/release")
+	if err = r.updateStatus(cfg, op.ConfigCondition{Code: op.AppliedPipelineVersion, Version: version}); err != nil {
+		log.Error(err, "failed to apply pipeline version")
+		return reconcile.Result{}, err
+	}
+
```
Comment (nikhil-thomas) [inline pkg/controller/config/config_controller.go:438]: shall we move this into `applyPipeline` or into `validatePipeline`

File: pkg/controller/config/config_controller.go [hunk 4]
```diff
@@ -231,9 +231,9 @@ func (r *ReconcileConfig) Reconcile(req reconcile.Request) (reconcile.Result, er
 		return r.applyPipeline(req, cfg)
 	case op.AppliedPipeline, op.PipelineValidateError:
 		return r.validatePipeline(req, cfg)
-	case op.ValidatedPipeline, op.AddonsError:
+	case op.ValidatedPipeline, op.AddonsError, op.AppliedPipelineVersion:
 		return r.applyAddons(req, cfg)
-	case op.AppliedAddons, op.CommunityResourcesError:
+	case op.AppliedAddons, op.CommunityResourcesError, op.AppliedTriggerVersion:
```
Comment (nikhil-thomas) [inline pkg/controller/config/config_controller.go:236]: could we reuse the existing states, `ValidatedPipelines` and `AppliedAddons` 
to handle the new additions

Comment (piyush-garg) [reply pkg/controller/config/config_controller.go:236]: we need status to add to the array

File: pkg/controller/config/config_controller.go (hunk 3 above)
Comment (piyush-garg) [reply pkg/controller/config/config_controller.go:438]: this is already in validatePipeline

**Comment by nikhil-thomas** [comment]
@piyush-garg 
the updated status structure in conditions list or (status message) will be a great context while reviewing this.
could you paste the config status yaml after the installation is successful (in the comment thread or in pr description)

**Comment by piyush-garg** [comment]
@nikhil-thomas 

Status will look like

```
[
  {
    "code": "installed",
    "version": "devel"
  },
  {
    "code": "trigger-version",
    "version": "v0.6.1"
  },
  {
    "code": "applied-addons",
    "version": "devel"
  },
  {
    "code": "pipeline-version",
    "version": "v0.14.2"
  },
  {
    "code": "validated-pipeline",
    "version": "devel"
  },
  {
    "code": "applied-pipeline",
    "version": "devel"
  }
]

```

**Comment by nikhil-thomas** [comment]
shall we separate the version as pipelines-version and triggers-version
something like
```
[
  {
    "code": "installed",
    "pipeline-version": "0.14.2"
    "triggers-version: "0.6.1"
  },
  {
    "code": "trigger-version",
    "pipeline-version": "0.14.2"
    "triggers-version: "0.6.1"
  },
  {
    "code": "applied-addons",
    "pipeline-version": "0.14.2"
    "triggers-version: ""
  },
  {
    "code": "pipeline-version",
    "pipeline-version": "0.14.2"
    "triggers-version: ""
  },
  {
    "code": "validated-pipeline",
    "pipeline-version": ""
    "triggers-version: ""
  },
  {
    "code": "applied-pipeline",
    "pipeline-version": ""
    "triggers-version: ""
  }
]
```

in addition, when we change the `conditions` to kubernetes style conditions later, the `pipeline-version` and `triggers-version` could become condition fields on `status`.

@vdemeester @sthaha

**Comment by piyush-garg** [comment]
@nikhil-thomas 

Now the status will look like

```
oc get config.operator.tekton.dev cluster -o json | jq '.status.conditions'
[
  {
    "code": "installed",
    "pipeline-version": "v0.14.3",
    "triggers-version": "v0.6.1",
    "version": "devel"
  },
  {
    "code": "applied-addons",
    "pipeline-version": "v0.14.3",
    "triggers-version": "v0.6.1",
    "version": "devel"
  },
  {
    "code": "validated-pipeline",
    "pipeline-version": "v0.14.3",
    "triggers-version": "",
    "version": "devel"
  },
  {
    "code": "applied-pipeline",
    "pipeline-version": "",
    "triggers-version": "",
    "version": "devel"
  }
]
```

File: pkg/controller/config/config_controller.go (hunk 4 above)
Comment (piyush-garg) [reply pkg/controller/config/config_controller.go]: removed the new status

**Comment by nikhil-thomas** [comment]
@vdemeester @sthaha

**Pull request merged**